File: frontend/src/lib/vietQr.ts
const DEFAULT_TEMPLATE = 'compact2';

const normalizeAmount = (amount: number): string => {
  const numeric = Number(amount ?? 0);
  if (Number.isNaN(numeric) || numeric < 0) {
    return '0';
  }
  return String(Math.floor(numeric));
};

const toEncodedValue = (value: string): string =>
  encodeURIComponent(String(value ?? '').trim());

interface BuildVietQrImageUrlParams {
  bankCode: string;
  accountNumber: string;
  template?: string;
  amount?: number;
  addInfo?: string;
  accountName?: string;
}

export const buildVietQrImageUrl = ({
  bankCode,
  accountNumber,
  template = DEFAULT_TEMPLATE,
  amount = 0,
  addInfo = '',
  accountName = '',
}: BuildVietQrImageUrlParams): string => {
  if (!bankCode || !accountNumber) {
    return '';
  }

  const path = `${toEncodedValue(bankCode)}-${toEncodedValue(accountNumber)}-${toEncodedValue(template)}.png`;
  const query = new URLSearchParams({
    amount: normalizeAmount(amount),
    addInfo: addInfo ?? '',
    accountName: accountName ?? '',
  });

  return `https://img.vietqr.io/image/${path}?${query.toString()}`;
};

interface BuildFromAccountBankParams {
  accountBank?: {
    bank?: string;
    accountNumber?: string;
    customerName?: string;
  } | null;
  amount?: number;
  addInfo?: string;
  template?: string;
}

export const buildVietQrImageUrlFromAccountBank = ({
  accountBank,
  amount = 0,
  addInfo = '',
  template = DEFAULT_TEMPLATE,
}: BuildFromAccountBankParams): string =>
  buildVietQrImageUrl({
    bankCode: accountBank?.bank ?? '',
    accountNumber: accountBank?.accountNumber ?? '',
    template,
    amount,
    addInfo,
    accountName: accountBank?.customerName ?? '',
  });

File: mobile/lib/utils/viet_qr.dart
const _defaultTemplate = 'compact2';

String _normalizeAmount(num amount) {
  final numeric = amount.toDouble();
  if (numeric.isNaN || numeric < 0) return '0';
  return numeric.floor().toString();
}

String _toEncodedValue(String value) =>
    Uri.encodeComponent(value.trim());

String buildVietQrImageUrl({
  required String bankCode,
  required String accountNumber,
  String template = _defaultTemplate,
  num amount = 0,
  String addInfo = '',
  String accountName = '',
}) {
  if (bankCode.isEmpty || accountNumber.isEmpty) return '';

  final path =
      '${_toEncodedValue(bankCode)}-${_toEncodedValue(accountNumber)}-${_toEncodedValue(template)}.png';
  final query = {
    'amount': _normalizeAmount(amount),
    'addInfo': addInfo,
    'accountName': accountName,
  };
  return 'https://img.vietqr.io/image/$path?${query.entries.map((e) => '${Uri.encodeQueryComponent(e.key)}=${Uri.encodeQueryComponent(e.value)}').join('&')}';
}

String buildVietQrImageUrlFromAccountBank({
  required String? bank,
  required String? accountNumber,
  required String? customerName,
  num amount = 0,
  String addInfo = '',
  String template = _defaultTemplate,
}) =>
    buildVietQrImageUrl(
      bankCode: bank ?? '',
      accountNumber: accountNumber ?? '',
      template: template,
      amount: amount,
      addInfo: addInfo,
      accountName: customerName ?? '',
    );
